crates/solidity/testing/sourcify/src/sourcify.rs

```diff
@@ -0,0 +1,323 @@
+            )),
+            should_save: self.should_save,
+        })
+    }
+
+    fn clean(&self) {
+        if !self.should_save {
+            fs::remove_dir_all(&self.path).unwrap();
+        }
+    }
+}
+
+impl Drop for Repository {
+    fn drop(&mut self) {
+        self.clean();
+    }
+}
+
+#[derive(Deserialize)]
+struct Manifest {
+    files: Vec<ManifestFile>,
+}
+
+impl Manifest {
+    fn get_chain_shards(&self, chain: Chain, options: &ShardingOptions) -> Vec<Shard> {
+        let mut shards: Vec<_> = self
+            .files
+            .iter()
+            .filter_map(|manifest| add_from_manifest(manifest, options, chain).ok())
+            .collect();
+
+        shards.sort_by(|a, b| a.id.cmp(&b.id));
+        shards
+    }
+}
+
+fn add_from_manifest(
+    file: &ManifestFile,
+    options: &ShardingOptions,
+    chain: Chain,
+) -> Result<Shard> {
+    // File path should come in this format:
+    // /sourcify-repository-2025-03-24T03-00-26/full_match.1.00.tar.gz
+    //                                                     - --
+    //                                                     | | shard prefix
+    //                                                     | chain ID
+    let mut parts = file.path.split('.');
+    let name_prefix = parts.next().unwrap();
+    let match_type = if name_prefix.ends_with("full_match") {
+        MatchType::Full
+    } else if name_prefix.ends_with("partial_match") {
+        MatchType::Partial
+    } else {
+        bail!("Invalid match type in archive path: {}", file.path);
+    };
+
+    if match_type == MatchType::Partial && options.exclude_partials {
+        bail!("Partials are excluded");
+    }
+
+    let chain_id_part = parts.next().ok_or(Error::msg("Failed to get chain ID"))?;
+    let chain_id: u64 = chain_id_part.parse()?;
+
+    if chain_id != chain.id() {
+        bail!("Wrong chain");
+    }
+
+    let id_part = parts
+        .next()
+        .ok_or(Error::msg("Failed to get shard prefix"))?;
+    let id = u16::from_str_radix(id_part, 16)?;
+
+    if !options.get_id_range().contains(&id) {
+        bail!("Shard not part of selected range.");
+    }
+
+    Ok(Shard {
+        path: file.path.clone(),
+        id,
+        match_type,
+    })
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct ManifestFile {
+    path: String,
+}
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum MatchType {
+    Full,
+    Partial,
+}
+
+impl MatchType {
+    pub fn dir_name(self) -> &'static str {
+        match self {
+            MatchType::Full => "full_match",
+            MatchType::Partial => "partial_match",
+        }
+    }
+}
+
+pub struct Shard {
+    pub id: u16,
+    pub match_type: MatchType,
+    /// A URL path used to fetch the `ContractArchive` for this shard.
+    pub path: String,
+}
+
+#[derive(Clone)]
+pub struct ContractArchive {
+    pub id: u16,
+    pub match_type: MatchType,
+    contracts_path: PathBuf,
+    should_save: bool,
+}
```

Comment:
There is a 1 to 1 relationship between these two, but it's unclear from the naming and API.